client/services.go

```diff
@@ -0,0 +1,154 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2017 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package client
+
+import (
+	"bufio"
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+
+	"github.com/snapcore/snapd/i18n"
+	"github.com/snapcore/snapd/systemd"
+)
+
+// ServicesOp encapsulate requests for performing an operation on a series of services
+type ServiceOp struct {
+	Services []string `json:"services,omitempty"`
+	Action   string   `json:"action"`
+}
+
+func (op ServiceOp) Description() string {
+	var verb string
+	switch op.Action {
+	case "enable-now":
+		verb = i18n.G("Enable and start")
+	case "disable-now":
+		verb = i18n.G("Stop and disable")
+	case "try-reload-or-restart":
+		verb = i18n.G("Try to reload or restart")
+	case "reload":
+		// the following are spelled out so xgettext finds them
+		verb = i18n.G("Reload")
+	case "start":
+		verb = i18n.G("Start")
+	case "stop":
+		verb = i18n.G("Stop")
+	case "enable":
+		verb = i18n.G("Enable")
+	case "disable":
+		verb = i18n.G("Disable")
+	case "restart":
+		verb = i18n.G("Restart")
+	default:
+		verb = strings.Title(op.Action)
+	}
+
+	if len(op.Services) == 0 {
+		// not currently supported
+		// TRANSLATORS: %s is the verb ("Stop", "Restart", etc)
+		return fmt.Sprintf(i18n.G("%s all services"), verb)
+	}
+
+	// TRANSLATORS: first %s is the verb ("Stop", "Restart", etc),
+	// second %s is list of service names ("a-snap.a-service, a-snap.b-service and %s")
+	tpl := i18n.NG("%s service %s", "%s services %s", uint32(len(op.Services)%1000000))
+
+	if len(op.Services) == 1 {
+		return fmt.Sprintf(tpl, verb, op.Services[0])
+	}
+
+	return fmt.Sprintf(tpl, verb,
+		// TRANSLATORS: first %s is a comma-separated list; second %s is the last element in the list
+		fmt.Sprintf(i18n.G("%s and %s"), strings.Join(op.Services[:len(op.Services)-1], ", "), op.Services[len(op.Services)-1]))
+
+}
+
+// A ServiceStatus is a description of a service's status in the system
+type ServiceStatus struct {
+	Snap    string `json:"snap"`
+	AppInfo        // note this is much less than snap.AppInfo, right now
+	*systemd.ServiceStatus
```

Comment:
This forces people to import snapd/systemd to be able to use the client API. We've been avoiding that sort of relationship in the past, so that one may import the client package without knowing about snapd specifics. We may even replace systemd by something else at some point and preserve the API here.

It's also quite awkward to have ServiceStatus.ServiceStatus as a field, and the note above makes me feel that AppInfo may be not be the right thing here as well.

What do we really want to report on a service status, and what's a clean way of representing it?

Considering these issues and this PR as a whole, I'm tempted to go back to the original point that we discussed, and attempting to make this API be about "apps" instead of about "services".

It'd be much nicer to have client.Logs, and AppInfo.Status, etc, instead of special casing services. We'll soon have timers as well, as systemd units. Are those services?  Why?  People will be able to alias them and run them in the command line just like everything else.

Let's please have a call tomorrow to discuss these details. It's really just mild changes on what is already in place, yet it unifies the terminology.